File: lexorder.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<long long> computeFactorials(int n) {
    vector<long long> factorials(n + 1, 1);
    for (int i = 2; i <= n; ++i) {
        factorials[i] = factorials[i - 1] * i;
    }
    return factorials;
}

vector<int> kthPermutation(int n, long long k) {
    vector<long long> factorials = computeFactorials(n);
    vector<int> available(n);
    vector<int> permutation;

    for (int i = 0; i < n; ++i) {
        available[i] = i + 1;
    }

    k--;  
    for (int i = 0; i < n; ++i) {
        long long base = factorials[n - 1 - i];
        int index = k / base;
        permutation.push_back(available[index]);
        available.erase(available.begin() + index);
        k %= base;
    }

    return permutation;
}

long long permutationToNumber(const vector<int>& perm) {
    int n = perm.size();
    vector<long long> factorials = computeFactorials(n);
    vector<bool> used(n + 1, false);
    long long number = 0;

    for (int i = 0; i < n; ++i) {
        int smaller = 0;
        for (int j = 1; j < perm[i]; ++j) {
            if (!used[j]) {
                smaller++;
            }
        }
        number += smaller * factorials[n - 1 - i];
        used[perm[i]] = true;
    }

    return number + 1;  
}

int main() {
    int n;
    long long k;
    cin >> n >> k;
    vector<int> before(n);
    for(int i = 0; i < n; i++) cin >> before[i];
    vector<int> permutation = kthPermutation(n, permutationToNumber(before) + k);
    for (int num : permutation) {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}

File: lexorder.py
def compute_factorials(n):
    factorials = [1] * (n + 1)
    for i in range(2, n + 1):
        factorials[i] = factorials[i - 1] * i
    return factorials

def kth_permutation(n, k):
    factorials = compute_factorials(n)
    available = list(range(1, n + 1))
    permutation = []

    k -= 1  
    for i in range(n):
        base = factorials[n - 1 - i]
        index = k // base
        if index >= available.__len__():
            from sys import exit
            print(-1)
            exit();
        permutation.append(available[index])
        available.pop(index)
        k %= base

    return permutation

def permutation_to_number(perm):
    n = len(perm)
    factorials = compute_factorials(n)
    used = [False] * (n + 1)
    number = 0

    for i in range(n):
        smaller = 0
        for j in range(1, perm[i]):
            if not used[j]:
                smaller += 1
        number += smaller * factorials[n - 1 - i]
        used[perm[i]] = True

    return number + 1  

if __name__ == "__main__":
    n, k = tuple(int(i) for i in input().split())
    before = list(map(int, input().split()))
    permutation = kth_permutation(n, permutation_to_number(before) + k)
    print(" ".join(map(str, permutation)))

